### PR TITLE
Apt history

### LINKED ISSUE #666: add support for advanced packaging tool (APT) history.log file
Add parser for apt logs. This log contains apt (advanced packaging tool) activity on a debian based system, e.g. installation/removal of packages.

Location: /var/log/apt/history.log

### PR DESCRIPTION
**Pull request by js-forensic**
## Description:
Added parser for Advanced Packaging Tool (APT) History log

**Related issue (if applicable):** fixes #666 

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/master/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

### CODE REVIEW

File: plaso/parsers/__init__.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from plaso.parsers import amcache
 from plaso.parsers import android_app_usage
 from plaso.parsers import apache_access
+from plaso.parsers import apthistory
```
Comment (Onager) [inline plaso/parsers/__init__.py]: Please rename to "apt_history" - and associated files. See "bash_history" for a similarly named parser/formatter etc.

File: plaso/parsers/apthistory.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  _APTHISTORY_LINE = (
+      pyparsing.ZeroOrMore(pyparsing.lineEnd()) +
+      pyparsing.Word('Start-Date: ').suppress() +
+      _APTHISTORY_DATE_TIME.setResultsName('start_date') +
+      pyparsing.SkipTo('End-Date: ').setResultsName('message') +
+      pyparsing.GoToColumn(10) +
+      _APTHISTORY_DATE_TIME.setResultsName('end_date') +
+      pyparsing.ZeroOrMore(pyparsing.lineEnd()))
+
+  LINE_STRUCTURES = [('logline', _APTHISTORY_LINE),]
```
Comment (Onager) [inline plaso/parsers/apthistory.py]: Remove the trailing comma from the list here

File: plaso/parsers/apthistory.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import errors
+from plaso.lib import definitions
+from plaso.parsers import logger
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+
+class AptHistoryLogEventData(events.EventData):
```
Comment (Onager) [inline plaso/parsers/apthistory.py]: It looks like the "commandline" string from the log would be useful to include - can you add this as well?

File: tests/parsers/apthistory.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  """Tests for the APT History log parser.
+
+  Since APT History logs record in local time, these tests assume that the local
+  timezone is set to UTC.
+  """
+
+  def testParseLog(self):
+    """Tests the Parse function on apt_history.log."""
+    parser = apthistory.AptHistoryLogParser()
+    storage_writer = self._ParseFile(['apt_history.log'], parser)
```
Comment (Onager) [inline tests/parsers/apthistory.py]: Please add checks for the "requestor" and "error" flags as well.

File: plaso/parsers/apthistory.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        event_data.requestor = line
+
+    for line in lines:
+      event_data.packages = line
+      if 'Downgrade:' in line:
+        event = time_events.DateTimeValuesEvent(
+            date_time,
+            definitions.TIME_DESCRIPTION_DOWNGRADE,
+            time_zone=time_zone)
+        parser_mediator.ProduceEventWithEventData(event, event_data)
```
Comment (Onager) [inline plaso/parsers/apthistory.py]: It looks like the idea is that only one of these strings appears in the line - if that's correct, please add a "continue" after the event and event_data are produced.

**Comment by Onager** [comment]
One other thing - please add this parser to the linux preset in data/presets.yaml

File: plaso/parsers/apthistory.py (hunk 5 above)
Comment (js-forensic) [reply plaso/parsers/apthistory.py]: This was deliberate. Not sure if it ever occurs, but my idea was that if more than one operation appears in a single log entry, I would create a separate event for each operation.

**Comment by js-forensic** [comment]
> Could you please refactor this to use the single line, rather than multiline parser? See #2718 for comparison.
> 
> Trying to parsing the entire entry at once can cause problems if the file is truncated, or the entry is exceptionally large.

Done

**Pull request merged**